crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs

```diff
@@ -319,6 +329,240 @@ async fn test_pinned_timeline_with_no_pinned_event_ids_is_just_empty() {
     test_helper.server.reset().await;
 }
 
+#[async_test]
+async fn test_edited_events_are_reflected_in_sync() {
+    let mut test_helper = TestHelper::new().await;
+    let room_id = test_helper.room_id.clone();
+
+    // Join the room
+    let _ = test_helper.setup_initial_sync_response().await;
+    test_helper.server.reset().await;
+
+    // Load initial timeline items: a text message and a `m.room.pinned_events` with
+    // event $1
+    let _ = test_helper
+        .setup_sync_response(vec![(msg("$1", "in the end"), false)], Some(vec!["$1"]))
+        .await;
+
+    let room = test_helper.client.get_room(&room_id).unwrap();
+    let timeline = Timeline::builder(&room)
+        .with_focus(TimelineFocus::PinnedEvents { max_events_to_load: 100 })
+        .build()
+        .await
+        .unwrap();
+    test_helper.server.reset().await;
+
+    assert!(
+        timeline.live_back_pagination_status().await.is_none(),
+        "there should be no live back-pagination status for a focused timeline"
+    );
+
+    // Load timeline items
+    let (items, mut timeline_stream) = timeline.subscribe().await;
+
+    assert_eq!(items.len(), 1 + 1); // event item + a day divider
+    assert!(items[0].is_day_divider());
+    assert_eq!(items[1].as_event().unwrap().content().as_message().unwrap().body(), "in the end");
+    assert_pending!(timeline_stream);
+    test_helper.server.reset().await;
+
+    // Load new pinned event contents from sync, where $2 is and edit on $1
+    let _ = Box::pin(
+        test_helper.setup_sync_response(vec![(edit("$2", "$1", "edited message!"), true)], None),
+    )
+    .await;
+
+    // The list is reloaded, so it's reset
+    assert_matches!(timeline_stream.next().await.unwrap(), VectorDiff::Clear);
```

Comment:
I think there's an `assert_next` that exists and could replace most of these `assert_matches` invokes here and below, and in the other test(s) too?

Reply:
For some reason changing `assert_matches!(timeline_stream.next().await.unwrap(), *)` to `assert_next_matches!(timeline_stream, *)` makes one of the tests fail with:

> assertion failed: stream is not ready, it's pending

Maybe I need to add a delay before? It's quite weird.

Reply:
Oh, it's because under the hood it calls `Stream::next_now_or_never`... That seems dangerous. I don't see any similar function for doing it while awaiting for the next item.

Reply:
We need to add a timeout then, otherwise if the stream item doesn't come immediately, then the test will be timing out with `cargo-nextest`'s delay, which is a minute or longer.

Reply:
Added a new `assert_next_matches_with_timeout!` macro for this. It's based on the `assert_next_matches`, but instead of checking the immediate result of the stream, it adds a small timeout which can be customised.